### ISSUE TITLE
openload.co hotlinking detection

### ISSUE THREAD

**Issue by bigmittens** [rg3/youtube-dl#9706]
### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2016.06.03*. If it's not read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2016.06.03**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through [README](https://github.com/rg3/youtube-dl/blob/master/README.md) and **most notably** [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones

### What is the purpose of your *issue*?
- [x] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add `-v` flag to **your command line** you run youtube-dl with, copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):
```
$ youtube-dl -v https://openload.co/f/kUEfGclsU9o
[debug] System config: []
[debug] User config: []
[debug] Command-line args: ['-v', 'https://openload.co/f/kUEfGclsU9o']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2016.06.03
[debug] Python version 3.5.1 - Linux-4.5.4-gnu-1-x86_64-with-arch
[debug] exe versions: ffmpeg 3.0.2, ffprobe 3.0.2, rtmpdump 2.4
[debug] Proxy map: {}
[Openload] kUEfGclsU9o: Downloading webpage
[debug] Invoking downloader on 'https://openload.co/stream/famBaQI9BRQ~1465252297~2602:306::~t47S-POR?mime=true'
[download] Destination: skyrim_no-audio_1080.mp4-kUEfGclsU9o.mp4
[download] 100% of 31.86MiB in 05:26
```

---

### Description of your *issue*, suggested solution and other information

openload.co no longer seeks to allow hotlinking and will redirect you to videos of pigeons instead of the video inputted to youtube-dl. This occurs on the latest version. Happens on both links like https://openload.co/f/kUEfGclsU9o and https://openload.co/embed/kUEfGclsU9o. I'm not familiar with the way the download link is returned, but it seems a solution would involve changing the method of requesting or returning the link.

**Comment by bigmittens** [comment]
Another thing: it would be nice for youtube-dl to pick up on subtitles/closed captioning from openload as well.

**Comment by reguez** [comment]
let's focus on the hotlinking detection for the moments

**Comment by arnrob14344** [comment]
youtube-dl -g -v https://openload.co/embed/JuMaadDghFY/
[debug] System config: []
[debug] User config: []
[debug] Command-line args: [u'-g', u'-v', u'https://openload.co/embed/JuMaadDghFY/']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2016.06.03
[debug] Python version 2.7.11 - Linux-4.5.4-1-ARCH-x86_64-with-glibc2.2.5
[debug] exe versions: ffmpeg 3.0.2, ffprobe 3.0.2, rtmpdump 2.4
[debug] Proxy map: {}
https://openload.co/stream/famBaQI9BRQ~1465428283~199.43.0.0~VS7nuSkm?mime=true
*************************
I get the link to the video but if i play it using mpv or other media player It shows only pigeons. Please help.

**Comment by reguez** [comment]
@arnrob14344  openload has recently implemented a hotlink detection , so you'r no longer able to get link directly unless you figure out how it work let's wait for an update from youtube-dl to solve this issue

**Comment by perron375** [comment]
ok they just move the script,
now is de 2nd;
I use this javascript regex "/<\/video>[\s\S]*?<script.*?>(.*?)<\/script>[\s\S]*?<script.*?>(.*?)<\/script>/)[2]"

**Comment by yokrysty** [comment]
They've made changes again, this time I get the pigeons video directly from their website player (without any hotlink)
I get 2 video links, the real one can be either of the 2 or sometimes neither (see below tests);
Those 2 links are saved in W array, the real video link is W[welikekodi_ya_rly] where welikekodi_ya_rly = 89 - 87; (random numbers every time)

I made a small script for testing that outputs the Content-Length for the 2 video files (33410733 is the size of the pigeons video):
```
C:\Users\KRYSTY>C:\Users\KRYSTY\Desktop\test.py
url: https://openload.co/embed/tdCcyArYqmA/
33410733
33410733
['https://openload.co/stream/tdCcyArYqmA~1465685183~2a02:2f09::~ksA9TlVg?mime=true', 'https://openload.co/stream/tdCcyArYqmA~1465685183~2a02:2f09::~ksA9TlVz?mime=true']

C:\Users\KRYSTY>C:\Users\KRYSTY\Desktop\test.py
url: https://openload.co/embed/tdCcyArYqmA/
33410733
33410733
['https://openload.co/stream/tdCcyArYqmA~1465685190~2a02:2f09::~9DdFwY-z?mime=true', 'https://openload.co/stream/tdCcyArYqmA~1465685190~2a02:2f09::~9DdFwY-1?mime=true']

C:\Users\KRYSTY>C:\Users\KRYSTY\Desktop\test.py
url: https://openload.co/embed/tdCcyArYqmA/
33410733
119914746
['https://openload.co/stream/tdCcyArYqmA~1465685209~2a02:2f09::~7DDInA8y?mime=true', 'https://openload.co/stream/tdCcyArYqmA~1465685209~2a02:2f09::~7DDInA8V?mime=true']

C:\Users\KRYSTY>C:\Users\KRYSTY\Desktop\test.py
url: https://openload.co/embed/tdCcyArYqmA/
33410733
33410733
['https://openload.co/stream/tdCcyArYqmA~1465685218~2a02:2f09::~_Y9XDtqE?mime=true', 'https://openload.co/stream/tdCcyArYqmA~1465685218~2a02:2f09::~_Y9XDtqX?mime=true']

C:\Users\KRYSTY>C:\Users\KRYSTY\Desktop\test.py
url: https://openload.co/embed/tdCcyArYqmA/
119914746
33410733
['https://openload.co/stream/tdCcyArYqmA~1465685224~2a02:2f09::~vKeBbfbr?mime=true', 'https://openload.co/stream/tdCcyArYqmA~1465685224~2a02:2f09::~vKeBbfb_?mime=true']
```

**Comment by perron375** [comment]
sorry but i dont know python hope this java script helps
ok i use this and works every time i tested many links
  //get the page
var data=get(l)

//get all scripts
    var aadecode_encoded = data.match(/<script.*?>(ﾟωﾟﾉ=[\s\S]*?)<\/script>/g);
// the result of welikekodi_ya_rly is allways 1 or 2 
    var welikekodi_ya_rly=new Function("return "+data.match(/welikekodi_ya_rly =(.*?;)/)[1])();

//here we get the good script using welikekodi_ya_rly
    var aascript= aadecode_encoded[welikekodi_ya_rly].split ("(ﾟДﾟ)[ﾟεﾟ]+");
end the rest is the same code

**Comment by sudovijay** [comment]
it look like it again broken. i can't extract link by this way

**Comment by arnrob14344** [comment]
same here i get this error:

$ youtube-dl -g -v https://openload.co/embed/Ggc850P9NWs
[debug] System config: []
[debug] User config: []
[debug] Command-line args: [u'-g', u'-v', u'https://openload.co/embed/Ggc850P9NWs']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2016.06.11.3
[debug] Python version 2.7.11 - Linux-4.5.4-1-ARCH-x86_64-with-glibc2.2.5
[debug] exe versions: ffmpeg 3.0.2, ffprobe 3.0.2, rtmpdump 2.4
[debug] Proxy map: {}
ERROR: Unable to extract video URL; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/youtube_dl/YoutubeDL.py", line 676, in extract_info
    ie_result = ie.extract(url)
  File "/usr/lib/python2.7/site-packages/youtube_dl/extractor/common.py", line 342, in extract
    return self._real_extract(url)
  File "/usr/lib/python2.7/site-packages/youtube_dl/extractor/openload.py", line 112, in _real_extract
    r'return\s+"(https?://[^"]+)"', decoded, 'video URL')
  File "/usr/lib/python2.7/site-packages/youtube_dl/extractor/common.py", line 645, in _search_regex
    raise RegexNotFoundError('Unable to extract %s' % _name)
RegexNotFoundError: Unable to extract video URL; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.

**Comment by sudovijay** [comment]
it just setting two global values i.e  window.M[1] & M[2]. which actually depend on the result of this expressions ! - welikekodi_ya_rly = a + b; , but sometime it do give pigeon on both links thats weird !

**Comment by JustMeDaFaq** [comment]
@sudovijay Encountering the same behaviour. Im parsing both links but sometimes both are pigeons. The result is totally random.

May theyre counting the video requsts from an ip and if its exceeds the allowed count it just results in 2 pigeon links?

**Comment by jakop345** [comment]
If the mime=true is removed at the end of the url it is working.

**Comment by sudovijay** [comment]
@jakop345  ! nope mime is there just for online stream. if you remove that file get force downloaded ! 

anyway i think the issue is resolved now as openload seem to be working fine for me.

**Comment by bigmittens** [comment]
@sudovijay still not working for me.

I don't even get the pigeons anymore, it just fails to extract.

**Comment by sudovijay** [comment]
@bigmittens ! ikr it just someone has to update the openload decoder!  it just youtube-dl just fetching the second js code "</video>\s*</div>\s*<script[^>]+>[^>]+</script>\s*<script[^>]+>([^<]+)</script>" 

Just solve the expression welikekodi_ya_rly. then decode the resulting js code. works every time ! alteast for now

**Comment by bigmittens** [comment]
@sudovijay ah, i thought you were saying the current release started working again. yeah, still waiting on the update. worked around it now by just going to the media tab on the info window on FF for the openload page and copying and pasting the video url into youtube-dl. 

hopefully we'll get a pull request soon!

**Comment by mariomosca** [comment]
Hi, i found this problem
ERROR: Unable to extract video URL; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 681, in extract_info
    ie_result = ie.extract(url)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 343, in extract
    return self._real_extract(url)
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/openload.py", line 112, in _real_extract
    r'return\s+"(https?://[^"]+)"', decoded, 'video URL')
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 646, in _search_regex
    raise RegexNotFoundError('Unable to extract %s' % _name)
RegexNotFoundError: Unable to extract video URL; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.

**Comment by perron375** [comment]
the var welikekodi_ya_rly is modified in the first aaencoded js code

**Comment by Belderak** [comment]
https://github.com/EPiC-APOC/repository.xvbmc/blob/a1fbaccb95df3866259bc81e9ec5fa0fa69228c7/script.module.urlresolver/lib/urlresolver/plugins/ol_gmu.py

Perhaps that could help

**Comment by yan12125** [comment]
This script is licensed under GPL3+, which is not compatible with youtube-dl (UNLICENSE), so we can't just copy and paste it.

**Comment by bigmittens** [comment]
is it me or has the site extraction method changed? now i cant even get the video without youtube-dl.

**Comment by yan12125** [comment]
The site has improved its protection so youtube-dl is broken for it.

**Comment by bigmittens** [comment]
@yan12125 we're talking about today/last night, yes? i was using the website fine last night and now it's completely broken for me. I can't watch any videos even in the browser. I'm currently tinkering around with add ons seeing if that could be the problem.

**Comment by yan12125** [comment]
Sorry I misunderstood you. Just tried https://openload.co/f/kUEfGclsU9o in Firefox and the video still plays.
Previous comments in this thread said the video player in the website is unstable. I guess it's what you have encountered.

**Comment by bigmittens** [comment]
I've fixed the issue in the browser. The problem was [CanvasBlocker](https://github.com/kkapsner/CanvasBlocker). It is on "fake readout API" which is usually harmless but it seems that the website now (as of yesterday) uses some sort of canvas fingerprint to display the video.

**Comment by gdkchan** [comment]
I spent the past few hours figuring out how the "obfuscation" javascript code on OpenLoad works, so I would like to share my findings here in the hopes that it will help with this issue. First, the page doesn't seems to be doing Canvas Fingerprinting, instead it seems to use the canvas to draw a png image from the page, and the colors on this image is later used to produce the stream url (this is why blocking the canvas will break the url generation). The meat of the url generation method is inside the javascript file "/assets/js/obfuscator/final.js", and is also uses data from "/assets/js/obfuscator/numbers.js". The page itself have a hidden base64 encoded PNG image with id "linkimg", and that image contains what at a first look may appear to be random noise, but the R, G and B channels on the color of each pixel encodes a ASCII character. This is used produce a string that is used on the final.js with the numbers.js string to generate the URL.

I'll not go into detail on that the javascript function actually does to generate the URL, since it can probably change in a near future anyway, but it's basically a code that uses data from both strings to generate another string, doing operations that doesn't really make much sense, even from a encryption view point.

I wrote a code in C# that takes a normal OpenLoad video URL and extracts the download URL, it is available [here](https://gist.github.com/gdkchan/2f88297203cd3189a41893e366faba8b).
Anyone is free to use the code.

I will need a few days to observate what the page will change on the decoding code to make it more future-proof.

Sorry for not directly contributing to youtube-dl but python really isn't my thing.

**Comment by bigmittens** [comment]
Interesting. Thank you for digging into the obfuscation and for the C# code. It's well written and seems to be fairly easy to understand. I'm sure this will be a big help for someone more well versed in python.

**Comment by yan12125** [comment]
@gdkchan It works! Can you declare your program as public domain so that I or someone else can translate it to Python and put it into youtube-dl?

Note to myself, maybe helpful for others: (How to run ```openload.cs``` under Linux)
```
mcs /reference:System.Drawing.dll openload.cs
mono openload.exe
```

**Comment by mrknow** [comment]
Here you have code to resolve openload links in  python. 
plugin IPTVPlayer  https://gitlab.com/iptvplayer-for-e2/iptvplayer-for-e2/blob/master/IPTVPlayer/libs/urlparser.py

**Comment by yan12125** [comment]
The script does not declare its license, so we can't use it, either. If you're the original author of that script, please consider declaring it as public domain or UNLICENSE.

**Comment by gdkchan** [comment]
Interesting, didn't know that someone else was working on this too. Anyway, I declared the code as public domain, so you're free to use or port it now.

**Comment by Belderak** [comment]
https://gist.github.com/Belderak/d891bba4ab06f08f6ec559adce182b73

Python code not clean at all but working...
Feel free to use.

**Comment by yan12125** [comment]
Thanks @Belderak I'll try to integrate this version. C# is too difficult for me :(

**Comment by Belderak** [comment]
Do you want some help ? I don't know if there's a helper for png data reading on youtube-dl

**Comment by yan12125** [comment]
youtube-dl has no support PNG. Are you going to implement a PNG decoder?

**Comment by Belderak** [comment]
If it has to be done, I will. Perhaps, there's already made class for that ?

**Comment by yan12125** [comment]
An WIP branch at: https://github.com/yan12125/youtube-dl/tree/openload-fix, based on @Belderak's approach. Need to find or write a pure Python implementation.

**Comment by yan12125** [comment]
Use pure Python [PyPNG](https://pythonhosted.org/pypng/) instead of pillow. To use it, run```pip install PyPNG```

Looks good on both Python 2 and Python 3. I'm not sure whether depending on external libraries is OK, though.

**Comment by Belderak** [comment]
Works on Win10 x64. But, I think we'll have to find a solution about the png.

**Comment by samsamsam-iptvplayer** [comment]
@Belderak
And you claim that you wrote this code: https://gist.github.com/Belderak/d891bba4ab06f08f6ec559adce182b73

by your self?
Without using my code:
https://gitlab.com/iptvplayer-for-e2/iptvplayer-for-e2/blob/master/IPTVPlayer/libs/urlparser.py

really? I don't think so. I wrote that at least you should write frome were you take the code.

**Comment by samsamsam-iptvplayer** [comment]
@yan12125
Only a blind man would not see which is this code.

**Comment by Belderak** [comment]
Nope. Sure, I used @gdkchan and your code to help me :> It's was just a gist to port something on Python. My code will not be on the git like that... I'll make a header to credit you.

**Comment by samsamsam-iptvplayer** [comment]
Sorry, but code @gdkchan is completely different. 
It's easy to spot that you simple take my code and use PLI.Image to decode PNG.
But why you remove this note:
"
# If you want to use the code for openload please at least put the info from were you take it:
# for example: "Code take from plugin IPTVPlayer: "https://gitlab.com/iptvplayer-for-e2/iptvplayer-for-e2/"
# It will be very nice if you send also email to me samsamsam@o2.pl and inform were this code will be used
"

Very bad behavior.

**Comment by yan12125** [comment]
> your code to help me

@Belderak Can you describe more? If any line in your code is copied from @samsamsam-iptvplayer's version, it's not clean. Those lines should be deleted and re-written.

**Comment by samsamsam-iptvplayer** [comment]
@yan12125 You really do not have eyes? 
When you change the variable name this is still the same code.

And as I wrote no problem to use this code. 
But why this notice about from where the code was take was removed?

**Comment by yan12125** [comment]
Let's compare the three versions:
@Belderak
```
width, height = img.size
output = ''
for y in range(0,height):
    for x in range(0,width):
        r, g, b = img.getpixel((x, y))
        if r==0 and g == 0 and b == 0:
            break
        else:
            output += str(unichr(r))
            output += str(unichr(g))
output += str(unichr(b))
```

**Comment by Belderak** [comment]
for i in [2, 3, 5, 7]:  
    linkData[i] = []
    Sum = 99
    for j in range(len(SigStr[i])):
        for k in range(len(SigStr[i][j])):
            if Sum > 122:
                Sum = ord('b')
            if SigStr[i][j][k] == unichr(int(math.floor(Sum))):
                if len(linkData[i]) > j:
                    continue
                Sum += 2.5;
                if k < len(ImgStr[i][j]):
                    linkData[i].append(ImgStr[i][j][k])

This is more taken from their than gdk's code
I'll rewrite it

**Comment by Belderak** [comment]
Gist updated rewritten like @gdkchan wrote it. 
Sorry for the inconvenience @samsamsam-iptvplayer

**Comment by yan12125** [comment]
Thanks @Belderak. My branch is updated as well. Just to mention there's a bug in your code:
```
res = 'https://openload.co/stream/' + Parts[3] + '~' + Parts[2] + '~' + Parts[1] + '~' + Parts[0]
```
By @gdkchan's version, it should be
```
res = 'https://openload.co/stream/' + Parts[3] + '~' + Parts[1] + '~' + Parts[2] + '~' + Parts[0]
```

**Comment by Belderak** [comment]
My bad. Sorry.

**Comment by yan12125** [comment]
Never mind. Thanks you all for tracking down this problem!

**Comment by samsamsam-iptvplayer** [comment]
It is simply ridiculous.

**Comment by yan12125** [comment]
What's your concern?

**Comment by Belderak** [comment]
Can't we merge the png.py and the LICENCE on youtube-dl ?
[](https://github.com/jcrocholl/pypng/blob/master/LICENSE)

**Comment by yokrysty** [comment]
instead of pyPNG, PIL library can be used:

```
def list2chunks(l, nth):
    return [l[x:x+nth] for x in range(0, len(l), nth)]
 
def string2chunks(s, n1, n2=0):
    chunks = map(''.join, zip(*[iter(s)]*n1))
    if n2 > 0:
        chunks = list2chunks(chunks, n2)
    return chunks
 
html = getHtml(get_embed_url(url))
 
if any(x in html for x in ['We are sorry', 'File not found']):
    raise Exception('The file was removed')
 
imageData = re.search(r'img id="linkimg" src="data:image\/png;base64,(.+?)"', html).group(1)
imageData = base64.b64decode(imageData)
 
#pyPNG
#rowst, columns, pngData, metadata = PNGReader(bytes=imageData).asDirect()
 
#PIL
pngData = Image.open(StringIO(imageData)).getdata()
 
signatureNumbers = getHtml('https://openload.co/assets/js/obfuscator/numbers.js')
signatureNumbers = re.search(r"window.signatureNumbers='(.+?)'", signatureNumbers).group(1)
 
imageStr = ''.join([chr(n) for n in itertools.chain.from_iterable(pngData) if n > 0])
imageData = string2chunks(imageStr, 20, 12)
signatureData = string2chunks(signatureNumbers, 26, 11)
 
videoUrlParts = []
for i in [2, 3, 5, 7]:
    s, n = '', ord('c')
    for j in range(len(signatureData[i])):
        for k in range(len(imageData[i][j])):
            if n > 122:
                n = ord('b')
            if signatureData[i][j][k] == chr(int(n)) and j >= len(s):
                n += 2.5;
                s += imageData[i][j][k];
    videoUrlParts.append(s.replace(',', ''));
 
videoUrl = 'https://openload.co/stream/%s?mime=true'
videoUrl = videoUrl % ('~'.join([videoUrlParts[i] for i in [3, 1, 2, 0]]))
 
print videoUrl
```

**Comment by gdkchan** [comment]
I just tried to run your code, and it shows this error:
`Traceback (most recent call last):
  File "openload.py", line 4, in <module>
    from PIL import Image
ImportError: No module named PIL`

Which means that PIL is not included with Python by default. And PyPNG also isn't. I don't know if using external libs is a problem for this project, but it seems to work with a default Python installation which is pretty nice. Shouldn't be that hard to make a PNG parser based on the specs, I think Python have ZLib support out of the box, so this is already half of the work done.

**Comment by yokrysty** [comment]
yeah it seems you are right, I installed python 2.7.12 in a virtual machine (fresh OS) and the PIL module is not present, so I am sure I installed it and I don't remember, my bad. I don't think the work on a new PNG library is worth the time, since openload  change the obfuscation technique quite often.

**Comment by yan12125** [comment]
The license of PyPNG is MIT-like. [1] I don't know whether or not there is a way to combine multiple licenses in a single software. Need a lawyer.

Technically PyPNG is a better choice than PIL. PyPNG is written in pure Python and PIL has C extensions. It's easier to include a copy of PyPNG in youtube-dl than another.

[1] MIT license template: https://opensource.org/licenses/MIT

**Comment by Belderak** [comment]
I think, we have to write our own PNG parser. Not a pro on that, but I can try.

**Comment by yokrysty** [comment]
I already started work to a simple png parser for this specific job :)

**Comment by Belderak** [comment]
Oh, keep going @yokrysty and good luck

**Comment by gdkchan** [comment]
I learned Python last night and made a quick png decoder that should work for the specific format used on OpenLoad images:

```
`def png2str(data):
	width = int(struct.unpack('>I', data[0x10:0x14])[0])
	height = int(struct.unpack('>I', data[0x14:0x18])[0])
	length = int(struct.unpack('>I', data[0x36:0x3a])[0])
	stride = width * 3
	
	comp = data[0x3e:0x3e + length]
	data = bytearray(zlib.decompress(comp))
	
	pxbuff = []
	
	str = ''
	
	for y in range(height):
		basePos = y * (1 + stride)
		filter = data[basePos]
		
		for x in range(stride):
			color = data[1 + basePos + x]
			basex = y * stride + x
			left = 0
			up = 0
			
			if (x > 2): left = pxbuff[basex - 3]
			if (y > 0): up = pxbuff[basex - stride]
			
			if filter == 1:
				color = (color + left) & 0xff
			elif filter == 2:
				color = (color + up) & 0xff
			elif filter == 3:
				color = (color + ((left + up) >> 1)) & 0xff
				#TODO: Paeth (not used by OL images it seems)
			
			str += chr(color)
			
			pxbuff.append(color)
			
		
			
	return str

**Comment by yokrysty** [comment]
@gdkchan nice job with the png parser it's simple and great for this job, no point to continue mine.

**Comment by yan12125** [comment]
@gdkchan That's amazing. Is this PNG parser also licensed as Public Domain or UNLICENSE?

**Comment by yan12125** [comment]
I'm not a lawyer. And I _guess_ verbally saying is enough.

**Comment by yokrysty** [comment]
@gdkchan made some modifcations to your code, hope you find them useful

```
def png2str(data):
    signature = struct.unpack('BBBBBBBB', data[:8])
    bytes = data[8:]
    
    if not signature == struct.pack('BBBBBBBB', 137, 80, 78, 71, 13, 10, 26, 10) and not bytes[4:8] == 'IHDR':
        raise IOError('Not a valid PNG file.')
    
    int_map = {1: '>B', 2: '>H', 4: '>I'}
    unpackInt = lambda x: struct.unpack(int_map[len(x)], x)[0]
    
    chunks = []
    
    while bytes:
        length = unpackInt(bytes[:4])
        bytes = bytes[4:]
        
        chunk_type = bytes[:4]
        bytes = bytes[4:]
        
        chunk_data = bytes[:length]
        bytes = bytes[length:]
        
        crc = unpackInt(bytes[:4])
        bytes = bytes[4:]
        
        chunks.append({
            'type': chunk_type,
            'length': length,
            'crc': crc,
            'data': chunk_data
        })
    
    ihdr = chunks[0]['data']
    
    width = unpackInt(ihdr[:4])
    height = unpackInt(ihdr[4:8])
    bitdepth = unpackInt(ihdr[8:9])
    colortype = unpackInt(ihdr[9:10])
    compression = unpackInt(ihdr[10:11])
    filtertype = unpackInt(ihdr[11:12])
    interlaced = unpackInt(ihdr[12:13])
    
    idat = ''
    
    for chunk in chunks:
        if chunk['type'] == 'IDAT':
            idat += chunk['data']
    
    if not idat:
        raise IOError('Unable to read PNG data.')
    
    data = bytearray(zlib.decompress(idat))
    
    str = ''
    
    stride = width * 3
    pxbuff = []

    for y in range(height):
        basePos = y * (1 + stride)
        filter = data[basePos]

        for x in range(stride):
            color = data[1 + basePos + x]
            basex = y * stride + x
            left = 0
            up = 0

            if (x > 2): left = pxbuff[basex - 3]
            if (y > 0): up = pxbuff[basex - stride]


            if filter == 1:
                color = (color + left) & 0xff
            elif filter == 2:
                color = (color + up) & 0xff
            elif filter == 3:
                color = (color + ((left + up) >> 1)) & 0xff
            elif filter == 4:
                a = left
                b = up
                c = 0

                if (x > 2 and y > 0): c = pxbuff[basex - stride - 3]

                p = a + b - c

                pa = abs(p - a)
                pb = abs(p - b)
                pc = abs(p - c)

                if pa <= pb and pa <= pc:
                    color = (color + a) & 0xff
                elif pb <= pc:
                    color = (color + b) & 0xff
                else:
                    color = (color + c) & 0xff

            str += chr(color)

            pxbuff.append(color)

    return str
```

**Comment by gdkchan** [comment]
@yokrysty Nice, it should be less sensitive to changes now.

**Comment by yokrysty** [comment]
I wonder how much time until they will change the obfuscation, since it's cracked once again :)

**Comment by gdkchan** [comment]
@yokrysty Well, I remember  making a decoder for the old AAEncode "protection" on february of this year, and someone else probably had figured it before me, and they just totally changed this one just recently, so this is half a year with broken protection. So on the best case scenario this one will last until next year at least. They will probably just do small changes to it to break plugins once they figure out it was cracked.

**Comment by yokrysty** [comment]
@gdkchan Yeah, I've made the first AAdecoder since I couldn't find any and I was in need for my Kodi. After a while there where other decoders, but mine has less code :) here take a look http://pastebin.com/4Dq9M86Y

**Comment by yokrysty** [comment]
openload video link extraction script http://pastebin.com/emMucDB8
credits goes to all of you here

**Comment by gdkchan** [comment]
Interesting. I made one based on the original encoder (doing the inverse process), and then made the Emoji table used on OL by hand. Then I noticed they was changed the emoji table once per week or something like that, breaking my decoder. So since the start of the javascript code was always the same, I made a method that used the beggining as a "guess window" to fill the emoji table, and after just decoded the URL, that was the only part that actually changed.

**Comment by yokrysty** [comment]
@gdkchan I've used some of the js interpreters (execjs, js2py, pyjsparser, this is what I used, there are more) for python in the past, but the problem is they get messed up with obfuscated code

**Comment by Belderak** [comment]
Nice job :D

**Comment by Belderak** [comment]
@yokrysty Can't we use the python wrapper of V8 engine ?

**Comment by Belderak** [comment]
Just need to wait @yan12125 PR so ?

**Comment by aarondiazr** [comment]
@yokrysty your script is awesome, but you need to change line 190 to:

`signatureNumbers = getHtml('https://openload.co/assets/js/obfuscator/n.js')
`
Because openload has change the name of "numbers.js"

**Comment by gdkchan** [comment]
Interesting, I updated my code with this change aswell.

**Comment by yan12125** [comment]
Things look good now. Will merge the fix into master soon.

**Comment by yokrysty** [comment]
off course they made a small useless change just to break our scripts

**Comment by Hrxn** [comment]
Since they are already here: Hello, openload guys! :cocktail:

**Comment by samsamsam-iptvplayer** [comment]
You really are funny.
You think you are the navel of the World?
I do not think so.

**Comment by Hrxn** [comment]
> You think you are the navel of the World?

Nobody is. Let's agree on that.

**Comment by samsamsam-iptvplayer** [comment]
So, why you think they are interested with this project? 
I do not think so. In my opinion they are interested about KODI plugins.

And youtube-dl is not project were "security" openload was broken, neither earlier nor now.

**Comment by yan12125** [comment]
Thanks everyone. Openload will be fixed in the next version.

@gdkchan @Belderak @yokrysty Would you like to be listed in [AUTHORS](https://github.com/rg3/youtube-dl/blob/master/AUTHORS)? If so under what name?

**Comment by gdkchan** [comment]
@yan12125 Nice, I don't think it's necessary but you can add me as "gdkchan" if you want.
I tested it and seems to be working good. The only weird thing i noticed is that it seems to append the video ID to the name. For example [this video](https://openload.co/f/AYgHe95d1E4/Sintel.2010.1080p.mkv.mp4) is saved as "Sintel.2010.1080p.mkv.mp4.mp4-AYgHe95d1E4.mp4" instead of "Sintel.2010.1080p.mkv.mp4". Don't know if this is intended through.

**Comment by bigmittens** [comment]
does anyone know if subtitles are obfuscated as well?

**Comment by gdkchan** [comment]
@bigmittens Do you have a example video with said subtitles?

**Comment by bigmittens** [comment]
@gdkchan I've thought I've seen videos with subtitles by default under the "CC", but now I'm looking for one with them to no avail. Maybe they were hardcoded all along. I'll let you know if I find any.

**Comment by samsamsam-iptvplayer** [comment]
@bigmittens 
Subtitles are not obfuscated. In IPTVPlayer I have implementation of extraction subtitles and they are working without any problems.

**Comment by bigmittens** [comment]
@samsamsam-iptvplayer do you have any openload links with subtitles?

**Comment by samsamsam-iptvplayer** [comment]
Yes, of course.

**Comment by bigmittens** [comment]
@samsamsam-iptvplayer could you post one, so that we could examine the method and make a cleanroom extraction?

**Comment by samsamsam-iptvplayer** [comment]
Here you are:
https://openload.co/embed/xZSf13ejvsM/Brief.Encounters.S01E01.720p.HDTV.x264-ORGANiC.mp4

**Comment by bigmittens** [comment]
thanks!

**Comment by samsamsam-iptvplayer** [comment]
No problem.

**Comment by yan12125** [comment]
> The only weird thing i noticed is that it seems to append the video ID to the name

@gdkchan By default youtube-dl adds video ID to filenames. If you don't like it, add ```-o "%(title)s.%(ext)s"``` to the command line.

**Comment by yan12125** [comment]
@bigmittens Please open a new issue for subtitles.

**Comment by bigmittens** [comment]
seems to be broken again

**Comment by gdkchan** [comment]
They seems to be using AAEncode again. The main page have 2 AAEncoded scripts. One was already present since the last version. The other one is the new URL decoding script. It takes the characters from a html element with ID "hiddenurl" and just deobfuscates it. It's more simple than the old method.

[Here](https://gist.github.com/gdkchan/10b1ab86fefe6940f4e5b7d9587c70b3) is some sample C# source to decode it.

Original JS code can be seen by AADecoding the first AAEncoded Javascript on the page (note that it have 2 inside the same tag, one to decode the link and the other to put the decoded link on the streamurl element using jQuery).

Curiously the linkimg image is still being served, just the old "obfuscator" code was removed. I guess that using the old numbers.js file would still work too.

**Comment by yan12125** [comment]
@gdkchan If I remember it correctly when you mentioned the PNG approach for the first time, there were still AAEncoded JavaScripts on their page? When I integrated the AAEncode-based algorithm in May (#8489), everything worked fine, and later they serve penguin videos or invalid files.

**Comment by gdkchan** [comment]
yes, it had an aaencoded javascript that was used to place the final stream
url on the streamurl tag. This script is still there with the other one
that does the decoding. both methods are working fine atm, but I guess they
will remove the png based one soon (as the reference to their respective
javascript files were removed from the page)

**Comment by yan12125** [comment]
OK. Just let me know then.

**Comment by bigmittens** [comment]
seems like the png method is gone

**Comment by yan12125** [comment]
Yes - there's a new issue at #10408. Please move there.

**Comment by ErkinAkbudak** [comment]
@yan12125  How to install openload-fix? Thank you.

**Comment by yan12125** [comment]
openload-fix was once working but now fails too. There's no known fix yet. Please move to #10408 to check the latest update or post your solution, if any.
